Speed-up discovery process

DmitrySamoylov opened the pull request:
I've left some comments in commit log, so it's better to review commit-by-commit.

DmitrySamoylov committed:
Disable tokio default features

DmitrySamoylov committed:
discovery: Filter-out duplicate WS-discovery responses

Sometimes we see duplicate responses coming from the network.

FSMaxB commented on onvif/Cargo.toml:
```diff
@@ -26,7 +26,7 @@ reqwest = { version = "0.11.0", default-features = false }
 sha1 = "0.6.0"
 thiserror = "1.0"
-tokio = { version = "1", features = ["sync", "time"] }
+tokio = { version = "1", default-features = false, features = ["net", "sync", "time"] }
```
Tokio (since 1.0) doesn't have any default features. Although `default-features = false` certainly doesn't hurt.

Moraxes commented on onvif/src/utils/display_list.rs:
```diff
@@ -0,0 +1,20 @@
+    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
+        formatter.write_char('[')?;
+
+        let mut peekable = self.0.iter().peekable();
+
+        while let Some(element) = peekable.next() {
+            element.fmt(formatter)?;
+            if peekable.peek().is_some() {
+                formatter.write_str(", ")?;
+            }
+        }
```
you can get rid of the peekable iterator with something like:

```rust
let mut is_first = true;
for element in self.0.iter() {
    if !is_first {
        formatter.write_str(", ")?;
    }
    is_first = false;
    element.fmt(formatter)?;
}
```

if you prefer. either way is fine for me tbh

DmitrySamoylov replied:
Yeah, it can also be `enumerate()` and then checking idx against `len()` but it's not very important, so will leave as is.

Moraxes commented on onvif/src/discovery/mod.rs:
```diff
@@ -115,9 +117,20 @@ pub async fn discover(duration: Duration) -> Result<impl Stream<Item = Device>,
-            debug!("Probe match XML: {}", xml);
+        while let Ok((xml, src)) = recv_string(&socket).await {
+            if !known_responses.insert(calculate_hash(&xml)) {
+                debug!("Duplicate response from {src}, skipping ...");
```
does a `warn!` make sense here? not sure how common this is, and whether it's a big deal or not

DmitrySamoylov replied:
Still needs investigation why this happens, but it's not critical at all.

Moraxes commented on onvif/src/discovery/mod.rs:
```diff
@@ -115,9 +117,20 @@ pub async fn discover(duration: Duration) -> Result<impl Stream<Item = Device>,
     let xml_receiver = ReceiverStream::new(xml_receiver);
     let device_receiver = ReceiverStream::new(device_receiver);
 
+    let mut known_responses = HashSet::new();
```
if you put this in the `produce_xmls` block, IMO you can insert entire XMLs instead of the hashes - this is technically worse wrt. memory usage (are the XMLs big?), but eliminates the tiny possibility of a hash collision, and memory usage is only increased for a brief time (the scope of the `produce_xmls` block). you could probably assume that the risk of duplicating XMLs only applies within a given `SocketAddr`, use a `HashMap<SocketAddr, String>`, then move the payload out of that `HashMap` when you're just about to `xml_sender.send(xml)`.

DmitrySamoylov replied:
> are the XMLs big?

I think around 1 KB, but who knows what the network device can send to us :smile: 

> memory usage is only increased for a brief time

It's for the full duration of discovery because we don't know when the duplicate arrives.

> tiny possibility of a hash collision

Even in case of collision, it's not that bad because the user can request discovery again, and it'll work the second time because there's randomness in the data.

Moraxes commented on onvif/src/discovery/mod.rs:
```diff
@@ -115,9 +117,20 @@ pub async fn discover(duration: Duration) -> Result<impl Stream<Item = Device>,
-        while let Ok(xml) = recv_string(&socket).await {
-            debug!("Probe match XML: {}", xml);
+        while let Ok((xml, src)) = recv_string(&socket).await {
+            if !known_responses.insert(calculate_hash(&xml)) {
```
if you choose to not go with the `HashMap` like I mentioned in the other comment, you should probably at least hash `(xml, src)` to correctly handle duplicate XMLs from different sources, but I'm not sure if there's any risk of that ever happening

DmitrySamoylov replied:
It'll be strange to have duplicates from different sources because XML contains IP addresses of the camera. Hashing `(xml, src)` will give the same amount of collisions, from what I can tell.

FSMaxB commented on onvif/src/discovery/mod.rs:
```diff
@@ -115,9 +131,20 @@ pub async fn discover(duration: Duration) -> Result<impl Stream<Item = Device>,
+                continue;
+            }
+
+            debug!(
+                "Probe match XML: {} Channel capacity: {}",
+                xml,
+                xml_sender.capacity()
+            );
```
Should the capacity and the XML be fields maybe?

DmitrySamoylov replied:
XML is too large for field, but capacity is good to put in a field :+1:

FSMaxB commented on onvif/src/discovery/mod.rs:
```diff
@@ -186,20 +215,51 @@ where
+                    debug!("{name:?}: Filtering out link-local IP address: {url}");
+                    false
+                }
+                // Filter out IPv6 addresses. Camera's usually have both, so let's use only IPv4.
```
You should account for cases where there is only an IPv6. I've seen this happen on some networks.

FSMaxB commented on onvif/src/discovery/mod.rs:
```diff
@@ -186,20 +215,51 @@ where
+
+            urls.retain(|url| match url.host() {
+                // Filter out link-local IP addresses.
+                // TODO: Make filtering link-local addresses configurable.
```
I wonder if we should just expose this entire filtering logic so an application can provide it's own. This means that we wouldn't have to touch `onvif-rs` everytime we need changes of this logic in `lumeod` and other users of the library can then also decide to do it differently.

DmitrySamoylov replied:
Yeah, good point. Probably, it's better to move most of this out of `onvif-rs`.

DmitrySamoylov committed:
discovery: Remove filtering discovered cameras

It can be undesirable for users to perform filtering
or they can choose another way to select a responding
URL.

DmitrySamoylov committed:
Add `utils::display_list`

We'll use it to debug-print `Vec<Url>` in a more readable way.

DmitrySamoylov committed:
discovery: Implement Debug for Device to make `urls` readable
